File: Dsa-day63.cpp
#include <bits/stdc++.h>
using namespace std;
class TaskManager {
    unordered_map<int,pair<int,int>> mp;
    priority_queue<tuple<int,int,int>> pq;
public:
    TaskManager(vector<vector<int>>& tasks){
        for(auto &x:tasks){
            mp[x[1]]={x[0],x[2]};
            pq.emplace(x[2],x[1],x[1]);
        }
    }
    void add(int userId,int taskId,int priority){
        mp[taskId]={userId,priority};
        pq.emplace(priority,taskId,taskId);
    }
    void edit(int taskId,int newPriority){
        int u=mp[taskId].first;
        mp[taskId]={u,newPriority};
        pq.emplace(newPriority,taskId,taskId);
    }
    void rmv(int taskId){
        mp.erase(taskId);
    }
    int execTop(){
        while(!pq.empty()){
            auto [p,tid,id]=pq.top();pq.pop();
            if(mp.count(tid)&&mp[tid].second==p){
                int u=mp[tid].first;
                mp.erase(tid);
                return u;
            }
        }
        return -1;
    }
};

File: Dsa-day63.py
import heapq

class TaskManager:
    def __init__(self, tasks):
        self.taskMap = {}
        self.heap = []
        for u,t,p in tasks:
            self.taskMap[t] = (u,p)
            heapq.heappush(self.heap, (-p,-t,t))
    def add(self, userId, taskId, priority):
        self.taskMap[taskId] = (userId,priority)
        heapq.heappush(self.heap, (-priority,-taskId,taskId))
    def edit(self, taskId, newPriority):
        u,_ = self.taskMap[taskId]
        self.taskMap[taskId] = (u,newPriority)
        heapq.heappush(self.heap, (-newPriority,-taskId,taskId))
    def rmv(self, taskId):
        if taskId in self.taskMap:
            del self.taskMap[taskId]
    def execTop(self):
        while self.heap:
            p,negT,t = heapq.heappop(self.heap)
            if t in self.taskMap and self.taskMap[t][1]==-p:
                u,_=self.taskMap.pop(t)
                return u
        return -1
